sdk/core/azure-core/azure/core/rest/_rest_py3.py

```diff
@@ -0,0 +1,522 @@
+    :ivar bool is_closed: Whether the network connection has been closed yet
+    :ivar bool is_stream_consumed: When getting a stream response, checks
+     whether the stream has been fully consumed
+    :ivar int num_bytes_downloaded: The number of bytes in your stream that
+     have been downloaded
+    """
+
+    async def read(self) -> bytes:
+        """Read the response's bytes into memory.
+
+        :return: The response's bytes
+        :rtype: bytes
+        """
+        if not self._has_content():
+            parts = []
+            async for part in self.iter_bytes():  # type: ignore
+                parts.append(part)
+            self._set_content(b"".join(parts))
+        return self._get_content()
+
+    async def iter_raw(self) -> AsyncIterator[bytes]:
+        """Asynchronously iterates over the response's bytes. Will not decompress in the process
+
+        :return: An async iterator of bytes from the response
+        :rtype: AsyncIterator[bytes]
+        """
+        # If you don't have a yield in an AsyncIterator function,
+        # mypy will think it's a coroutine
+        # see here https://github.com/python/mypy/issues/5385#issuecomment-407281656
+        # So, adding this weird yield thing
+        for _ in []:
+            yield _
+        raise NotImplementedError()
+
+    async def iter_bytes(self) -> AsyncIterator[bytes]:
+        """Asynchronously iterates over the response's bytes. Will decompress in the process
+
+        :return: An async iterator of bytes from the response
+        :rtype: AsyncIterator[bytes]
+        """
+        # If you don't have a yield in an AsyncIterator function,
+        # mypy will think it's a coroutine
+        # see here https://github.com/python/mypy/issues/5385#issuecomment-407281656
+        # So, adding this weird yield thing
+        for _ in []:
+            yield _
+        raise NotImplementedError()
+
+    async def iter_text(self) -> AsyncIterator[str]:
+        """Asynchronously iterates over the text in the response.
+
+        :return: An async iterator of string. Each string chunk will be a text from the response
+        :rtype: AsyncIterator[str]
+        """
+        async for byte in self.iter_bytes():  # type: ignore
+            text = byte.decode(self.encoding or "utf-8")
+            yield text
+
+    async def iter_lines(self) -> AsyncIterator[str]:
+        """Asynchronously iterates over the lines in the response.
+
+        :return: An async iterator of string. Each string chunk will be a line from the response
+        :rtype: AsyncIterator[str]
+        """
+        async for text in self.iter_text():
+            lines = parse_lines_from_text(text)
+            for line in lines:
+                yield line
+
+    async def close(self) -> None:
+        """Close the response.
+
+        :return: None
+        :rtype: None
+        """
+        self.is_closed = True
+        self.internal_response.close()
```

Comment:
This is never an async operation?

Reply:
yeah, that's bc the internal responses's `close` s are not awaitable. You're comment does make me think that in the base class, it can be neater and we just call `await self.internal_response.close()`, and I'll move this onto the base classes themselves. Thanks!